### ISSUE TITLE
Do not inhibit backups and auto-saves

### ISSUE THREAD

**Issue by phil-s** [josteink/emacs-oob-reboot#25]
These settings added in https://github.com/josteink/emacs-oob-reboot/pull/7

Autosaves and backup files are there to *protect* the user. Disabling that protection is a bad idea IMO. What's the rationale behind doing that?

Similarly for create-lockfiles.

I don't think disabling any of the protective features in Emacs makes for sensible default behaviour.

**Comment by ssunday** [comment]
@phil-s If this feature of Emacs was better documented and explained to the user, then I might agree–as something for the user to turn on. As it stands, for a new-user, having files clutter up their directory and autosaves happen without their clear knowledge is contrary to workflow in other text editors. Other popular editors do not create extra files and periodically save by default. It also has the potential for the polluting of a project in VCS as the new user may not realize they are tracking these files.

Basically, it should be default off and then the user can choose to enable this protection with the knowledge of how to actually use it. Personally, and anyone I know who uses Emacs, doesn't use these features.

**Comment by ssunday** [comment]
@phil-s Expanding: if this is to make OOB Emacs experience less painful, disabling features that are very Emacs specific and require a more intermediate knowledge make sense. When I first started Emacs, I was like why are these files being created? What are they? Why is the file saving automatically? I had to search to figure that out. It wasn't obvious.

But yes, maybe it helps the user, but only a more _advanced_ user. Not a user that just was introduced to Emacs.

**Comment by brotzeit** [comment]
What is the default of other editors ? I also think this is a feature, users usually activate after they decided to give a new editor a try.

**Comment by bretfh** [comment]
Auto saving is a feature that is best when its out of the way. Not much
needs to be said on what auto save means for any user.

Emacs mode line even periodically says its saving the file. How is this not
enough?

Focusing on what other editors are doing isnt the best way to go about
this. Inspiration perhaps yes but calling some feature as not modern or
user friendly solely because another editor doesnt have it means nothing.

Besides its hard to say what is user friendly and what is not. People come
from different backgrounds with different concepts of all things.


Emacs is best when the user is allowed to shape it in the way they wish.
This is the power and beauty of emacs. ITs not what we agree emacs should
be. Its how we tailor make it for ourselves. So instead of focusing on
rather small details on how a defualt emacs should look why not instead put
users in a position to explore and shape emacs to their liking? A better
means of exploration, documentation, tutorials, etc. Improve the means for
the user to know emacs and shape emacs to their liking.

**Comment by brotzeit** [comment]
> So instead of focusing on rather small details on how a defualt emacs should look why not instead put users in a position to explore and shape emacs to their liking?

I totally agree, but most users need time to feel comfortable enough to even think about building their own environment.

**Comment by dchrzanowski** [comment]
I feel like the clutter is completely unnecessary. If a user wants it they can google it or be told about it in the tutorial, apart from that I say no to the backups. Auto-save is absolutely fine in my opinion.

**Comment by angrybacon** [comment]
@Grimscythe1983 It's hard to google for something you have no clue might exist. Also not everyone reads tutorials in full.

**Comment by dchrzanowski** [comment]
@angrybacon are you then suggesting that we should leave the backup feature on?

**Comment by angrybacon** [comment]
@Grimscythe1983 Yes absolutely, both backup files and lock files. And if we agree to leaving the backup files, then leave auto-save on for sure too; if not, perhaps increase the value for `auto-save-timeout`.

**Comment by dchrzanowski** [comment]
Ok then, can I then suggest to put all the backup files into a single directory, so that the feature doesn't trash?

**Comment by danlamanna** [comment]
> Ok then, can I then suggest to put all the backup files into a single directory, so that the feature doesn't trash?

This is what I was going to say, put them in `~/.emacs.d/tmp` or something out of the way.

**Comment by dchrzanowski** [comment]
Something like this `(setq backup-directory-alist '(("." . "~/.emacs.d/backup")))` does the trick.

**Comment by angrybacon** [comment]
What's the harm in leaving them next to their related file? It's so much easier to access them. If the issue is Git-related (or similar), I found it to be a good opportunity to setup my global `.gitignore`.

**Comment by josteink** [comment]
> Yes absolutely, both backup files and lock files. And if we agree to leaving the backup files, then leave auto-save on for sure too; if not, perhaps increase the value for auto-save-timeout.

As someone who considers myself a advanced user by now, I've had nothing but trouble with the lock-files. I find them highly unintuitive, and I am consistently annoyed whenever I meet them. The only thing I see them doing, is breaking my workflow.

File-locking, if done (and that can be debated. No other editor on the market locks files!), should be done at OS-level, with real file-locks, provided by the OS.

What Emacs does doesn't actually work the second you step outside the Emacs-frame, and even inside Emacs it feels broken.

The current implementation in Emacs is terrible, and I really think it would benefit new users to disable it, maybe coupled with `auto-revert-mode`, to avoid conflicts when a file is open/edited in multiple editors or programs at the same time.

Here's a clear vote for disabling locking.

**Comment by dchrzanowski** [comment]
> What's the harm in leaving them next to their related file? 

Tidiness? I personally hate clutter, I simply hate when files show up that I haven't made myself.

> Here's a clear vote for disabling locking.

Same here, disable locking. It is very confusing for a new user.

**Comment by ssunday** [comment]
@josteink Completely agree.

**Comment by pkkm** [comment]
Backups are an useful safety feature and have saved me from losing a couple hours of work more than once. I think we should consider solutions other than disabling them altogether, e.g. introducing a size limit and ``(setq backup-directory-alist `(("." . ,(locate-user-emacs-file "backups"))))``, or perhaps using a per-drive and per-user directories (similar to how Linux file managers handle the Trash folder).

**Comment by ssunday** [comment]
@pkkm How long did it take you to figure out how to use it? (I'm asking because I don't think a new user can utilize it with any effectiveness)

**Comment by angrybacon** [comment]
It doesn't take much to understand what they are, they are grayed out and visiting them should make it clear that they are an older version of the file.

**Comment by ssunday** [comment]
@angrybacon True, but it _does_ increase the bar for people to get into Emacs.

**Comment by pkkm** [comment]
I don't remember exactly, but I think I figured it out quickly. I might be an outlier though, I was already a programmer and had used another editor which makes backups.

Anyway, I agree that the default backup settings aren't intuitive for new users; my point is that there may be a good solution that doesn't litter all over the place but still makes some backups. New users may not even know about the backups until they accidentally overwrite something important and google for ways to recover it. I don't have an exact proposal, but I think the two possible solutions I described in my previous post are a good start for discussion.

**Comment by alandmoore** [comment]
>What's the harm in leaving them next to their related file?

Contrived example:  Imagine a piece of software that automatically includes all config files in a directory.  I edit a file in that directory.  Now I have both my new edited version and the old version, both of which the program includes because the programmer forgot to exclude files ending in ~. Kaboom!

I have my autosaves consigned to something like ~/.emacs.d/backups.  The downside here is that it silently fills up over time and could potentially be a security problem (e.g., I edit a root-owned credentials file with mode 600 on a remote server using tramp, now the backup save is in my home directory, owned by me.  oops....).

I don't know what the right course of action is here, but I'd agree that the current default is not usual or expected.

**Comment by angrybacon** [comment]
> Now I have both my new edited version and the old version, both of which the program includes because the programmer forgot to exclude files ending in ~.

That requires 1. that the user forgot / didn't know he had to ignore backup files, 2. that he added the backup files to the commit (himself, not Emacs on its own) and 3. that the user decides to commit configuration files at all. Usually I see `.sample` files.

> Kaboom!

Huh?

> now the backup save is in my home directory, owned by me

Yeah that could be bad. Why not use Emacs from that remote instead? Or better disable backups altogether when doing remote?

**Comment by alandmoore** [comment]
@angrybacon I think you're losing sight of my point.  We're not talking about me, we're talking about a new user, who has only used more "conventional" editors in the past, and is trying out emacs.

- The config file example was a bit contrived, and probably very unlikely; but things like this have happened to me in the past when working with less-than brilliantly written software.  Yes, there are things that "could have" or "should have" been done, the point is that people unused to the emacs way of working wouldn't necessarily expect that save file to be there, and in such a situation it can cause problems.  And unless you `ls` your directory after every save, you're not likely to notice these files (or know where they came from).

- With the tramp/remote thing -- the point is that if autosave is enabled by default and directed to a hidden directory, inexperienced users won't know of its existence and will likely make this exact blunder.  

>Why not use Emacs from that remote instead?

Perhaps Emacs isn't installed on the remote?  Certainly not with my personal settings.  

>Or better disable backups altogether when doing remote?

Why would a new user do this if they didn't even know backups were being done in the first place?  Heck, I don't even know how to temporarily disable backups, and I wouldn't remember to do it every time I edited a potentially sensitive file.

I don't know what the answer is on the autosave issue.  I think that:

- The current default behavior is a useful safety feature, at least for config files and random editing (One would hope for codebases you're using version control).  Leaving those save files everywhere is kind of a nuisance.
- Disabling it puts emacs on par with most other editors in this regard.  It removes a significant annoyance, but at the cost of a potentially butt-saving feature.
- Redirecting the files to a hidden location is probably a bad idea.  It's only a good idea if you know it's being done and know how to manage it.  For beginners, not such a good idea.

**Comment by angrybacon** [comment]
> the point is that people unused to the emacs way of working wouldn't necessarily expect that save file to be there

I'm assuming the user is familiar with basic usage of a version-control system. The user would have to deliberately add the backup files himself.

> Perhaps Emacs isn't installed on the remote? Certainly not with my personal settings.

That's the thing I was implying using your own configuration remotely after installing whichever version suits you. But I get it, not always an option.

>  inexperienced users won't know of its existence and will likely make this exact blunder

Well, we're talking about a user who attempts to work on a remote machine where he knows root-owned credentials live from his local Emacs. I'd expect that person knows what he's doing. But you're also right, he wouldn't automatically guess that backups are made at all.

> Redirecting the files to a hidden location is probably a bad idea. It's only a good idea if you know it's being done

One way of ensuring they know backup files are being saved somewhere would be not to make it a default behavior. :-)

**Comment by amlamarra** [comment]
Whatever you all decide on, one thing I can recommend is that when closing emacs (or just the buffer) without saving changes, emacs should ask if you want to save the file only ONCE. One thing that constantly bothered me was it asking me twice and STILL leaving those backup files. To me, if I intentionally close without saving, it should remove those backup files. Or am I wrong on this? I thought the point was to protect you from system (or just emacs) crashes. I made a change in my config to only prompt once, though I don't have access to it at the moment to show what I'm talking about. And I support having all those files with # and ~ (I forget which one does which) saved to a single location in ~/.emacs.d

**Comment by angrybacon** [comment]
> it should remove those backup files

Do you mean lock files? Backup files, lock files and auto-save are 3 different things.

**Comment by amlamarra** [comment]
> Do you mean lock files? Backup files, lock files and auto-save are 3 different things.

I'm referring to the backup files (~) & auto-save files (#). I'm not exactly sure why Emacs uses both. I understand the usefulness of the auto-save files and I don't seem to have a problem with them, they go away if emacs is properly closed. The backup files, however, stay. Why?

**Comment by angrybacon** [comment]
When you have different save slots in a game, they don't disappear when you exit the game. :-)

**Comment by amlamarra** [comment]
I don't think that's the best analogy. The "game" here is a text editor and the save slots are you text files. Why is my game creating hidden backups of my save slots? I understand this may have been useful when Emacs was first created, when hard drives (and floppy drives) were less reliable and the chance of file corruption was much greater. But today... me thinks it no so necessary.

**Comment by angrybacon** [comment]
This is regardless of where we should store the backup files.

In my analogy, the save slots of your game are `foo.el~`. The auto-save slot (the one you load when you just select "Continue" as opposed to "Load") is `foo.el`, it's the latest version. Provided the game supports the "Load" feature and allows to keep several slots of saves, then I wouldn't expect it (the game) to just delete those previous saves (`foo.el~`) upon restart.

**Comment by amlamarra** [comment]
Again, that's not the best analogy. When I intentionally save a game, of course I don't expect it to delete those upon restart. In this analogy, that would be your foo.el file. foo.el is not the game, it's your save file. The 'game' is needlessly creating a duplicate of your save file. Well, needlessly to me. Unless someone can provide a good reason for this.

**Comment by phil-s** [comment]
All backups are completely useless -- until the time comes when you need one.  You'll be amazed at just how grateful you are for them when that day comes.

**Comment by amlamarra** [comment]
Sure, but if every program made duplicate copies of your saved files for reasons of "just in case," our hard drives would be cluttered with crap. It's on the user to backup their own files. Most programmers use github (or something similar) and a local VCS. Personally, I also use a cloud backup solution. I used to keep everything in Dropbox until I setup my own Nextcloud server and use that now. I'd hate it if every program did what Emacs does by default. Emacs isn't even very good at doing what it does. Isn't the auto-save file (starts & ends with the # character) supposed to be automatically deleted when Emacs is properly closed? I can't tell you how many times that doesn't happen. There's a line between protecting the users from themselves & just being annoying. Prompting to close emacs (or the buffer) and loosing unsaved changes is, obviously, perfectly acceptable. But the default behavior of creating multiple copies of the same file in the same location is just annoying. Though I'm also fine with just keeping all those extra copies in a specific location (such as somewhere in the ~/.emacs.d directory).

**Comment by dchrzanowski** [comment]
@amlamarra I just couldn't agree more!

**Comment by angrybacon** [comment]
Version control and backup are 2 very different things.

**Comment by amlamarra** [comment]
Agreed. But if you push your changes to github (or some other remote repo), now you've got a backup of your repo in case your hard drive fails or something.

**Comment by angrybacon** [comment]
My point is version control and backup files treat 2 different issues. Version control keeps a history, backup files are a security.

**Edit**: "Security" as in, you can't expect the user to know in advance when he is going to experience a major issue resulting in file loss.

**Comment by dchrzanowski** [comment]
@angrybacon you can't deny the fact that distributed version control systems provide a backup, its more like two birds with one stone - it is a history and it is a backup. Isn't this one of the purposes of GitHub and places of similar nature?

**Comment by angrybacon** [comment]
@Grimscythe1983 Sure, I agree that you can use Git to store your backups and build a history with your *backups* (which in reality are versions). Github "and places of similar nature" however have an entirely different purpose, ie. provide online storage for these versions, promote contributions/sharing and social interactions (Github is Git + hub).

Using Git or any other version control system as a primary backup is just asking for trouble. That is because your editor is better suited for it: it's directly between you and your work, it knows when a backup should be made. On the other hand, Git, for instance, has to be told when to make the "backup" which makes it unreliable. It's reliable until it isn't.

**Comment by dchrzanowski** [comment]
Hmmm... I do not agree with your statement. But hey! Freedom of speech, right?

**Comment by angrybacon** [comment]
How will Git save you in the case of power outage? Or more frequently, when you lose the connection over SSH. Both these examples are not properly dealt with using Git as a mean for backup.

**Comment by josteink** [comment]
> Most programmers use github (or something similar) and a local VCS

We shouldn't assume all Emacs-users are programmers. That said...

> Sure, but if every program made duplicate copies of your saved files for reasons of "just in case," our hard drives would be cluttered with crap. It's on the user to backup their own files.

Right now only Emacs does this. And some may say that it's a good thing, a Emacs-feature, one which sets it apart from the competition and which we should both use and market.

But I don't think that argument holds. If every other editor out there copied this "good feature", it would seriously become annoying. So how can it be a good feature if it loses value the more software implements it?

I too vote we disable these backups.

**Comment by jackrusher** [comment]
I understand that some of you have never used this feature, but it is one of the very nice things that emacs does. I vote against disabling it.

**Comment by ssunday** [comment]
@josteink Agreed completely. Well, I agree with all arguments in favor of disabling it as I was the one who originally disabled it on the repo haha.

**Comment by justinjk007** [comment]
@jackrusher yes they have saved me number of times, I actually have this bit

```emacs-lisp
;;Put backup files neatly away -- saved me many times
(let ((backup-dir "~/Dropbox/Code/Emacs/backups")
      (auto-saves-dir "~/Dropbox/Code/Emacs/autosavedir/")
      )
  (dolist (dir (list backup-dir auto-saves-dir))
    (when (not (file-directory-p dir))
      (make-directory dir t)))
  (setq backup-directory-alist `(("." . ,backup-dir))
        auto-save-file-name-transforms `((".*" ,auto-saves-dir))
        auto-save-list-file-prefix (concat auto-saves-dir ".saves-")
        tramp-backup-directory-alist `((".*" . ,backup-dir))
        tramp-auto-save-directory auto-saves-dir))
(setq backup-by-copying t    ; Don't delink hardlinks
      delete-old-versions t  ; Clean up the backups
      version-control t      ; Use version numbers on backups,
      kept-new-versions 3    ; keep some new versions
      kept-old-versions 2)   ; and some old ones, too
```
in my init file to store all the backups and autosaves into a separate Directory.

**Comment by jackrusher** [comment]
@justinjk007 I do it like this :) 
```
;; place backup files here, rather than sprinkling them everywhere.
(setq backup-directory-alist
      `((".*" . ,(expand-file-name
                  (concat user-emacs-directory "backups")))))
(setq auto-save-file-name-transforms
      `((".*" ,temporary-file-directory t)))
```

**Comment by alandmoore** [comment]
I think it's fairly clear that any backup can potentially be quite useful and save the day.

I think it's also fairly clear that this particular feature is implemented in a way that is unconventional, surprising, and (apparently, considering how many alter/disable it) annoying.

So the operative question is, which one matters more?  There isn't a lot of consensus on that here, though we are only a small group of users.

**Comment by phil-s** [comment]
I see it like this...

Emacs makes back-ups automatically, which is awesome because, as it was just put, those backups can save the day.

Disabling backups is bad, as then the day cannot be saved.

Writing the backups to a separate directory is really useful (IMO; it's certainly what I do), but could be seen as a negative if users aren't aware that it's happening (i.e. it's nice to be *aware* that your software is making copies of files).

The default behaviour makes it *really obvious* that the backup copies are being made, because they're right there in the same directory.

Users initially won't know what those files are, however *it's incredible easy to do a web search to find out what the files are and why they are being created.*

People performing such searches typically want to know how to prevent the files from being created, at which point it is typically explained to them (a) how to do what they asked, and (b) how to do what they *didn't* ask -- make the backups go somewhere different, so they're still being made, but no longer causing the visible clutter that was bothering them.

So that brings us back to writing backups to a separate directory as a solution, except *now they know it's happening, and where the backups live* which I think is pretty important.

In other words I think the backups should remain enabled, and that the best way for people to *learn about them* is probably to retain the default behaviour.

I think any elisp customisation around this issue should be squarely targeted at streamlining that learning process -- helping users to understand about backups/autosaves/etc, so they can make their own informed decision.

**Comment by amlamarra** [comment]
I can totally see that perspective, especially from a new-user point of view. Whatever default is decided, maybe this is something that could be configured at initial setup. I think I saw discussion elsewhere of having Emacs ask the user some questions for its initial configuration. This seems like something that should be asked of the user. "Do you want Emacs to keep backup copies of your files?" If yes, "Where would like them stored?" That sort of thing.

**Comment by alandmoore** [comment]
>I think I saw discussion elsewhere of having Emacs ask the user some questions for its initial configuration.

I feel like people are saying this as a way around emacs having opinionated defaults, or of settling a contentious issue; but I personally think it's a bad idea in general to bombard new users with a lot of questions about preferences whose consequences they can't really appreciate.  How obnoxious would it be to play 20 questions every time I run emacs as a different user on a different server or workstation instance?

**Comment by leppert** [comment]
> I feel like people are saying this as a way around emacs having opinionated defaults

👍🏻

**Comment by ssunday** [comment]
@alandmoore If I was bombarded with 20 questions every time I installed and started emacs on all the servers I need to do work on, I wouldn't bother with emacs anymore. I'd tough it out through vi/nano.

We should have sensible defaults that make enough sense that we don't have to ask a question at install time.

**Comment by phil-s** [comment]
a) I imagine any such questions could be opt-in, behind some kind of "Configure defaults" link on the splash screen.

b) Surely you install your emacs config whenever you install emacs??

**Comment by ssunday** [comment]
@phil-s No, not if its one of many servers I've spun up and the only reason I'm installing emacs is to quickly edit a few files. The server/user aren't necessarily exclusively mine so it wouldn't be good to install MY emacs config on something that isn't solely mine.

**Comment by angrybacon** [comment]
Those questions could be asked when Emacs doesn't find an init file. Behind one first prompt "Looks like this is your first day in Emacs, do you want to configure things? [Y/n]" with better wording.

**Comment by justinjk007** [comment]
@ssunday Can't we just use tramp for that sort of thing? I forgot the last time I had install emacs...I remember spending a good ten minutes installing build tools and grabbing dependencies is that not needed theses days or can you just install it straight away ?

**Comment by ssunday** [comment]
@justinjk007 On Centos I can just `yum install emacs`...takes maybe a minute.

**Comment by dchrzanowski** [comment]
I recently worked on Arch and on Mint / Ubuntu, Emacs package is  precompiled in the repositories, so installation is a walk in the park.

**Comment by alandmoore** [comment]
I suppose a setup wizard could conceivable be done in a non-obnoxious way, and might deal with some of the more contentious issues (cua-mode, e.g.) in a way that's actually helpful; but unless we have a way to commit code to this project to activate or implement a setup wizard, it's kind of a moot point.  

It seems to me, though, that perhaps emacs-devel might like to know that certain existing defaults are contentious to a significant number of users, even if they're desirable defaults to enough that we can't form a consensus to change it.  For instance, with this issue, it seems equally incorrect to say "users are happy with backups as-is" (which we'd be saying if we committed no code) and "users want backups disabled" (which we'd be saying with the original PR).  Maybe @phillord could comment on how (or if) we would most effectively proceed on contentious issues.  Maybe a separate "contentious" branch?  

Or, maybe I'm overthinking this.

**Comment by justinjk007** [comment]
Yes +1 for setup wizard from me, and @ssunday for bypassing the setup wizard we could easily setup a keyboard shortcut which will inhibit setup wizard writing it to the init file. 

I should also mention no other editors or IDEs do this, I mean the setup wizard, but hey emacs is more than that, so we should.

I really wish I would have known about CUA mode when I started, I mean I know it is at the end of the Guided Tour, if the devs think people will read all the way there to use CUA-mode then implemented a setup-wizard is the way to go IMHO.

**Comment by josteink** [comment]
I think recent changes to Emacs (eglot, tree-sitter, etc) has helped make Emacs much more usable OOB. Considering repo-purpose obsolete.

Closing all issues and archiving repo. Thanks to all contributors!
